### PR TITLE
[WIP] OAuth app registration

### PR DESCRIPTION
**Pull request by abought**
## Purpose
1. Provide models and UI to allowing users to register new OAuth2 applications
2. Demonstrate the first front-end views built on the new API. This entails adding support for cross-origin requests to the API and javascript calls.

## Changes
1. Add a new database model (OAuth2App) to track information about client 
2. Add new APIv2 endpoints that allow users to request and change information about their API applications (default intended restriction: only registered owner can make queries about an application)
3. Add new UI at `/settings/applications` and `/settings/applications/<client_id>` allowing users to see list of currently registered applications, view detailed information, or create new.
4. Add (and configure) the django-cors-headers application, used by DRF to support OSF calls to the new API endpoints

## To be done
- [x] Tests, o' glorious tests
- [ ] UI / knockout cleanup to make consistent with OSF standards
- [ ] Add styling to display of client id/ client secret on detail page
- [x] Address initial user feedback/ review
- [x] Move "viewer must be owner" API checks into a separate permissions class, rather than hardcoding inside views

Possibly post-MVP: 
- [ ] Add ability to revoke tokens etc via calls to CAS
- [ ] Allow resetting of client secret via UI. May also invalidate tokens as part of this reset.

## References
- Issue #30

### CODE REVIEW

File: api/users/serializers.py [hunk 1]
```diff
@@ -50,3 +68,68 @@ class ContributorSerializer(UserSerializer):
     filterable_fields = frozenset.union(UserSerializer.filterable_fields, local_filterable)
 
     bibliographic = ser.BooleanField(help_text='Whether the user will be included in citations for this node or not')
+
+    def get_bibliographic(self, obj):
```
Comment (brianjgeiger) [inline api/users/serializers.py]: This seems out of scope for this PR. Also, I expect this is being handled by the view. Is this something you put in, or is it just a side effect of some merges?

File: api/users/serializers.py [hunk 2]
```diff
@@ -50,3 +68,68 @@ class ContributorSerializer(UserSerializer):
...
+        node = self.context['view'].get_node()
+        return obj._id in node.visible_contributor_ids
+
+
+class OAuth2AppSerializer(JSONAPISerializer):
+    """Serialize data about a registered OAuth2 application"""
```
Comment (brianjgeiger) [inline api/users/serializers.py]: Have we decided if we want to make this internal-only or not? If we do, we'll want to ensure that the docs routes don't display this information.

File: api/users/serializers.py (hunk 1 above)
Comment (abought) [inline api/users/serializers.py]: I... have no memory of this line- possibly a side effect of some merges. Thanks for catching that.

File: api/users/serializers.py (hunk 2 above)
Comment (abought) [inline api/users/serializers.py]: > Have we decided if we want to make this internal-only or not? If we do, we'll want to ensure that the docs routes don't display this information.

There's no good reason I can think of to publicize the existence of this endpoint. @icereval would have the final say.

File: api/users/serializers.py [hunk 3]
```diff
@@ -50,3 +68,68 @@ class ContributorSerializer(UserSerializer):
...
+                          validators=[user_validator])
+
+    name = ser.CharField(help_text="A short, descriptive name for this application",
+                         required=True)
+
+    description = ser.CharField(help_text="An optional description displayed to all users of this application",
+                                required=False,
+                                allow_blank=True)
+
+    reg_date = ser.DateTimeField(help_text="The date this application was generated (automatically filled in)",
```
Comment (brianjgeiger) [inline api/users/serializers.py]: Prefer whole words in variable names. Also, we kind of overuse 'registration' on the OSF, so maybe a different noun there.

File: api/users/urls.py [hunk 4]
```diff
@@ -5,4 +5,6 @@
     url(r'^$', views.UserList.as_view(), name='user-list'),
     url(r'^(?P<pk>\w+)/$', views.UserDetail.as_view(), name='user-detail'),
     url(r'^(?P<pk>\w+)/nodes/$', views.UserNodes.as_view(), name='user-nodes'),
+    url(r'^(?P<pk>\w+)/applications/$', views.ApplicationList.as_view(), name='application-list'),
+    url(r'^(?P<pk>\w+)/applications/(?P<client_id>\w+)/$', views.ApplicationDetail.as_view(), name='application-detail')
```
Comment (brianjgeiger) [inline api/users/urls.py]: I think we are currently breaking these off into their own modules and having the modules handle sub-routing. I know currently it would be one route, but it's probably best to follow the pattern.

File: api/users/views.py [hunk 5]
```diff
@@ -86,3 +87,80 @@ def get_queryset(self):
...
+    permission_classes = (
+        drf_permissions.IsAuthenticated,
+    )
+
+    serializer_class = OAuth2AppSerializer
+
+    # overrides RetrieveAPIView
+    def get_object(self):
+        id_url_kwarg = 'client_id'
+        query = Q('client_id', 'eq', self.kwargs[id_url_kwarg])
```
Comment (brianjgeiger) [inline api/users/views.py]: Super-minor: This might be a little easier to understand without the variable. `query = Q('client_id', 'eq', self.kwargs['client_id'])` since it doesn't seem like you're reusing the variable.

File: api/users/views.py [hunk 6]
```diff
@@ -86,3 +87,80 @@ def get_queryset(self):
...
+        obj = get_object_or_404(OAuth2App, query)
+
+        # TODO: Write test for case where user is not logged in, or wrong user requests resource
+        if obj.owner._id != self.request.user._id:
```
Comment (brianjgeiger) [inline api/users/views.py]: This should be handled by a permissions class. I think @billyhunt is writing one for his work, so wait on this change until his stuff is merged in.

File: website/profile/views.py [hunk 7]
```diff
@@ -25,8 +27,9 @@
 from website import mails
 from website import mailchimp_utils
 from website import settings
-from website.models import User
-from website.models import ApiKey
+from website.models import (User,
```
Comment (brianjgeiger) [inline website/profile/views.py]: This should all fit on one line without the parentheses, I believe. #superMinor

File: website/profile/views.py [hunk 8]
```diff
@@ -356,6 +359,40 @@ def user_notifications(auth, **kwargs):
         'mailing_lists': auth.user.mailing_lists
     }
 
+@must_be_logged_in
+def oauth_application_config(auth, **kwargs):
+    """Return app creation page with list of known apps"""
+    # TODO: Hardcoded URL
+
+    url_string = "v2/users/{}/applications/".format(auth.user._id)
+    app_list_url = urljoin(settings.API_DOMAIN, url_string)
```
Comment (brianjgeiger) [inline website/profile/views.py]: I know you know this, but from the django settings also join API_PATH and API_BASE after settings.API_DOMAIN. Then take off the v2 from url_string.

File: website/profile/views.py [hunk 9]
```diff
@@ -356,6 +359,40 @@ def user_notifications(auth, **kwargs):
...
+    """Register an API application: blank form view"""
+    url_string = "v2/users/{}/applications/".format(auth.user._id)  # POST request to this url
+    submit_url = urljoin(settings.API_DOMAIN, url_string)  # TODO: hardcoded url
+    return {"submit_url": json.dumps(submit_url),
+            "detail_url": json.dumps(None)}
+
+@must_be_logged_in
+def oauth_application_detail(auth, **kwargs):
+    """Show detail for a single OAuth application"""
+    client_id = kwargs.get('cid')
```
Comment (brianjgeiger) [inline website/profile/views.py]: Per the newly made https://github.com/brianjgeiger/osf.io/issues/96 let's go ahead and make that `client_id` instead of `cid`.

File: website/profile/views.py (hunk 8 above)
Comment (abought) [inline website/profile/views.py]: Thanks for the reminder- that's an important detail!

File: website/profile/views.py [hunk 10]
```diff
@@ -356,6 +359,40 @@ def user_notifications(auth, **kwargs):
...
+
+    # TODO: This method requires 2 database hits (page load + API call) instead of 1. Is there a more concise way?
```
Comment (brianjgeiger) [inline website/profile/views.py]: You could include applications in the user object. If you do, I'd do it for other reasons and not just to simplify this one step. This won't be hit enough to make it necessary for a speed optimization, and it's fairly clear from a flow standpoint, so probably not worth getting clever on.

File: website/project/model.py [hunk 11]
```diff
@@ -285,6 +286,50 @@ def node(self):
...
+    # TODO: Rename to is_active and update serializers + michael CAS code
+    active = fields.BooleanField(default=True)  # Set to False if application is deactivated for users
+
+    owner = fields.ForeignField('user', backref='created')
+
+    # User-specified application descriptors
+    name = fields.StringField(index=True, required=True)
+    description = fields.StringField(required=False)
+
+    reg_date = fields.DateTimeField(default=datetime.datetime.utcnow, required=True)
```
Comment (brianjgeiger) [inline website/project/model.py]: Same as above

File: website/project/model.py [hunk 12]
```diff
@@ -285,6 +286,50 @@ def node(self):
...
+
+    home_url = fields.StringField(required=True)
```
Comment (brianjgeiger) [inline website/project/model.py]: Is this what I suggested we use for CORS? If so, you might want to allow this to handle multiple URLs.

Comment (brianjgeiger) [inline website/project/model.py]: Even if we don't have it do so in the front end yet, if we might need it to do so in the future, an array here would probably save a migration later.

Comment (abought) [inline website/project/model.py]: Let's talk to @icereval first: there are many use cases in which the application is being run on the back-end, not from within a web browser. 

I'm not 100% clear that opening up the CORS whitelist is the best way to give remote applications access (and it might have larger implications than we expect).

I think the OAuth2 provider authentication class may provide a completely separate access option that doesn't depend on CORS, but I'd like to get feedback.

Comment (brianjgeiger) [inline website/project/model.py]: :+1:

File: website/static/js/apiApplication.js [hunk 13]
```diff
@@ -0,0 +1,323 @@
...
+    self.reg_date = ko.observable();
+
+    self.owner = ko.observable();
+    self.name = ko.observable();
+    self.description = ko.observable();
+    self.homeUrl = ko.observable();
+    self.callbackUrl = ko.observable();
+
+    // Unchanging properties
+    self.detailUrl = ko.observable();
```
Comment (brianjgeiger) [inline website/static/js/apiApplication.js]: If these are unchanging, they won't need to be observable, they can just be variables.

File: website/static/js/apiApplication.js [hunk 14]
```diff
@@ -0,0 +1,323 @@
...
+    }
+    };
+
+// Load data from JSON
+ApplicationData.prototype.fromJSON = function (data) {
+    var self = this;
+    data = data || {};
+    self.clientId(data.client_id);
+    self.clientSecret(data.client_secret);
+    self.reg_date(data.reg_date);
```
Comment (brianjgeiger) [inline website/static/js/apiApplication.js]: And here as well.

File: website/static/js/apiApplication.js [hunk 15]
```diff
@@ -0,0 +1,323 @@
...
+    $.ajax({
+        type: 'GET',
+        url: url,
+        dataType: 'json',
+        // Enable CORS
+        crossOrigin: true,
+        xhrFields: {
+            withCredentials: true
+        },
+        success: function (data) {
```
Comment (brianjgeiger) [inline website/static/js/apiApplication.js]: Prefer promises to callbacks. https://cosdev.readthedocs.org/en/latest/style_guides/javascript.html#promises

Also, I expect a lot of your DRY concerns in the todos will be made simpler by passing in a function that's defined elsewhere rather than doing things inline.

**Comment by brianjgeiger** [comment]
Okay, first pass of WIP done. Some variables, some minor things, and a lot of promises should keep you busy for a while. :penguin:

**Comment by abought** [comment]
I make no promises!

But I do want to use them. Agreed; a lot of this will be made more concise by implementing osfhelpers functions that can handle CORS. The existing request helpers in that library could do with a revamp anyway.

File: api/users/urls.py (hunk 4 above)
Comment (abought) [inline api/users/urls.py]: These routes live under users by design ("must be owner to see data on this application"). What would be the best way to break this into a separate module while still living under /users ?

`v2/users/<user_id>/applications/<app_id>`

Comment (brianjgeiger) [inline api/users/urls.py]: I'm not positive. Maybe a module api.users.applications?

Comment (brianjgeiger) [inline api/users/urls.py]: However, that being said, we're not really doing that for nodes. I was probably thinking of that wrong. Let's leave them where they are for now.

File: api/users/views.py (hunk 6 above)
Comment (abought) [inline api/users/views.py]: > This should be handled by a permissions class. I think @billyhunt is writing one for his work, so wait on this change until his stuff is merged in.

What's the tracking issue for the relevant work? Not seeing any new activity in `api/nodes/permissions.py`.

Comment (brianjgeiger) [inline api/users/views.py]: Here's the PR. https://github.com/brianjgeiger/osf.io/pull/94

Comment (abought) [inline api/users/views.py]: > Here's the PR. #94

Thanks. Though given that this endpoint serves up client secret keys, I'm not sure that `ReadOnlyOrCurrentUser` is sufficient: if the request doesn't come from the app owner, I don't think we should be serving up *any* information. 

Just blocking updates seems insufficient... but then, I've missed the obvious before. :)

Comment (brianjgeiger) [inline api/users/views.py]: Nope, you're correct. Go ahead and make a new permissions class.

**Comment by brianjgeiger** [comment]
When you get this to the next stage of wanting to be reviewed, close this and make it a feature PR to the main repo's develop.

**Comment by abought** [comment]
Moved WIP PR to main repo.
